### ISSUE TITLE
[browser] Support AppContext switches on runtime

### ISSUE THREAD

**Issue by maraf** [dotnet/runtime#97449]
AppContext switches defined in csproj are not available on runtime

```xml
<ItemGroup>
    ...
    <RuntimeHostConfigurationOption Include="System.Runtime.OptionalFeatureBehavior"
                                    Condition="'$(UnoptimizedUserFacingBehavior)' != ''"
                                    Value="$(UnoptimizedUserFacingBehavior)"
                                    Link="false" />
    ...
</ItemGroup>
```

On other platforms, they end up in `runtime.config.json`. We should put them into `blazor.boot.json` as that's the only config required (for both wasmbrowser and blazor experience)

**Comment by lewing** [comment]
> AppContext switches defined in csproj are not available on runtime
> 
> ```
> <ItemGroup>
>     ...
>     <RuntimeHostConfigurationOption Include="System.Runtime.OptionalFeatureBehavior"
>                                     Condition="'$(UnoptimizedUserFacingBehavior)' != ''"
>                                     Value="$(UnoptimizedUserFacingBehavior)"
>                                     Link="false" />
>     ...
> </ItemGroup>
> ```
> 
> On other platforms, they end up in `runtime.config.json`. We should put them into `blazor.boot.json` as that's the only config required (for both wasmbrowser and blazor experience)

That works for browser but we will have the same issue in wasi where there is no json parser.

**Comment by maraf** [comment]
> That works for browser but we will have the same issue in wasi where there is no json parser.

That's true. We can use the binary format for WASI and we bundle it into wasm file even for browser, but I would like to avoid distributing it as a separate file for browser scenario

**Comment by lambdageek** [comment]
Mono has a task that converts rutnimeconfig.json files into a binary blob at build time so you don't need to do JSON parsing at execution time 
https://github.com/dotnet/runtime/blob/main/src/tasks/MonoTargetsTasks/RuntimeConfigParser/RuntimeConfigParser.cs

the binary blob can be passed by the driver to `monovm_runtimeconfig_initialize`:

https://github.com/dotnet/runtime/blob/77fd98c8fb245b30b24b8eeaf577b0a457a4449c/src/native/public/mono/jit/details/mono-private-unstable-functions.h#L21

https://github.com/dotnet/runtime/blob/77fd98c8fb245b30b24b8eeaf577b0a457a4449c/src/native/public/mono/jit/details/mono-private-unstable-types.h#L21-L34

This is used by xamarin-android, for example:

https://github.com/xamarin/xamarin-android/blob/e67afc7d7ae848da6b33026b80973d30713b98ad/src/monodroid/jni/monodroid-glue.cc#L846-L848

**Comment by am11** [comment]
FWIW, Native AOT passes `RuntimeHostConfigurationOption` to ilc during the publish time https://github.com/dotnet/runtime/blob/ef1a8ab967065da279479f7dfe67d9b778a50d29/src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets#L267 which materializes the key/value list as `g_compilerEmbeddedKnobsBlob` in the binary. At run-time, it populates a dictionary with those during init: https://github.com/dotnet/runtime/blob/cb29b0864583d70718d13c6679ff32608d2ca8e9/src/coreclr/nativeaot/System.Private.CoreLib/src/System/AppContext.NativeAot.cs#L15
called here https://github.com/dotnet/runtime/blob/a9dcd70f312b97e8e0b558dd3df2ffbccc720319/src/libraries/System.Private.CoreLib/src/System/AppContext.cs#L18-L20

We could use the same approach in wasm/wasi, i.e. write them as C#, typescript or C file under obj/ dir in pre-publish step and let it compile as part of the output object. The idea is the initial values (RuntimeHostConfigurationOption) are immutable/frozen between publish and run. During run-time AppContext.SetData etc. allows us to overwrite the value because it's just a regular str->obj dictionary. So if the use-case is to be able to manually change the values between publish and run, we can drop it because NAOT (and probably singlefile apps) also doesn't allow that type of thing; modification of value require republishing the app.

**Comment by pavelsavara** [comment]
> it's just a regular str->obj dictionary.

I wonder why this is not str->str in `AppContext`. 
Maybe you can add objects when the app is running ?
But the `Setup` only deals with string values.

> We could use the same approach in wasm/wasi, i.e. write them as C file

Yes, this is probably the right approach for WASI. (which is out of scope for this issue)

In browser case, we also need this to work without LLVM and so the easiest way is to add it to `dotnet.boot.js` (formerly known as blazor.boot.json) have some exported `AddKeyValue(string k, string v)` and call it from TS.

**Comment by am11** [comment]
> I wonder why this is not str->str in `AppContext`.

`AppContext.SetSwitch` and `AppContext.TryGetSwitch` expect `bool`s. 
`AppContext.SetData` and `AppContext.GetData` expect `object`s.

I think this shape makes more sense in the legacy context when multiple AppDomains were supported and it was there to share the in-memory data (SetData/GetData) in domain context during the process lifetime. The relationship with configs wasn't evolved in .NET Framework as it is today. In .NET Core, since there is only one AppDomain, the in-memory use-case isn't used much. Now its role is limited to be a low-level configuration system (mechanics are a bit involved 😅) https://learn.microsoft.com/dotnet/fundamentals/runtime-libraries/system-appcontext.

**Comment by radekdoulik** [comment]
We can also put the binary blob to a data section with a new wa-info based task. Then it can be used to initialize a range of memory. This would work for both browser and wasi. (or we can use emcc's --embed-file option on browser and find similar solution for wasi)

Optionally it can be also added as a custom section, which can be accessed from JS on browser. On wasi it would not work though I guess.
